File: preprocess/dataArgumentAndCrop.m

%% Clean
clear
close all

%% Parameters
global batchSize cropWidth widthNeighbor
global frameExt nArguments argumentZoom
alignments = {'_nowarp'};%{'_nowarp','_OF','_homography'};
inputDir = '/home/zhangcheng/projects/DeepVideoDeblurring/data';
inputFolderPrefix = 'training_real_all_nostab';
gtDir = '/home/zhangcheng/projects/DeepVideoDeblurring/dataset/quantitative_datasets';

saveDir = '../data';
saveFolderPrefixTrain = 'training_augumented_all_nostab';
saveFolderPrefixValid = 'validating_augumented_all_nostab';

frameExt = '.jpg';
cropWidth = 128;
nCrop = 10;
nArguments = 2*4*4;
widthNeighbor = 5;
batchSize = 64;
currSpeed = 0;
argumentZoom = [1/4, 1/3, 1/2, 1];

%% Scan videos
inputFolders = {};
for alignment = alignments
    inputFolders{end+1} = [inputFolderPrefix,alignment{1}];
end
inputFolders = fullfile(inputDir,inputFolders);
videoNames = {};
for inputFolder = inputFolders
    videoNames{end+1} = dir(inputFolder{1});
    maskFolders = [videoNames{end}.isdir];
    videoNames{end} = videoNames{end}(maskFolders);
    videoNames{end} = videoNames{end}(3:end);
    videoNames{end} = {videoNames{end}.name};
end

validset = {'IMG_0030' 'IMG_0049' 'IMG_0021' '720p_240fps_2' 'IMG_0032' ...
'IMG_0033' 'IMG_0031' 'IMG_0003' 'IMG_0039' 'IMG_0037'};% from git rep DeepVideoDeburring

%% Generate batches
for iAlignment = 1:length(alignments)
    alignment = alignments{iAlignment};
    
    %% scan frames
    trainset = setdiff(videoNames{iAlignment},validset);
    frameDirs = scanFrames(inputFolders{iAlignment},gtDir,trainset);
    
    %% generate batches
    saveAlignFolder = fullfile(saveDir,[saveFolderPrefixTrain,alignment]);
    checkDir(saveAlignFolder);
    generateBatches(frameDirs,saveAlignFolder,nCrop*length(frameDirs)/batchSize*nArguments)
    
    %% scan validset frames
    frameDirs = scanFrames(inputFolders{iAlignment},gtDir,validset);
    
    %% generate validset batches
    saveAlignFolder = fullfile(saveDir,[saveFolderPrefixValid,alignment]);
    checkDir(saveAlignFolder);
    generateBatches(frameDirs,saveAlignFolder,nCrop*length(frameDirs)/batchSize*nArguments)
end


%% functions 

function frameDirs = scanFrames(inputFolder,gtDir,videoNames)
global frameExt widthNeighbor
% for all videos
disp('Scanning frames...');
frameDirs = {};
iFrameAll = 1;
for iVideo = 1:length(videoNames)
    videoName = videoNames{iVideo};
    inputFrameFolder = fullfile(inputFolder,videoName);
    inputFrameNames = dir(fullfile(inputFrameFolder,'image_0',['*',frameExt]));
    inputFrameNames = {inputFrameNames.name};
    GTFrameFolder = fullfile(gtDir,videoName,'GT');
    GTFrameNames = dir(fullfile(GTFrameFolder,['*',frameExt]));
    GTFrameNames = {GTFrameNames.name};
    for iFrame = 1:length(inputFrameNames)
        neighborDirs = cell(widthNeighbor,1);
        for iNeighbor = 1:widthNeighbor
            iFrameNeighbor = iNeighbor-ceil(widthNeighbor/2);
            neighborDirs{iNeighbor} = [inputFrameFolder,'/',sprintf('image_%d',iFrameNeighbor),'/',inputFrameNames{iFrame}];
        end
        frameDirs{iFrameAll,1} = neighborDirs;
        frameDirs{iFrameAll,2} = [GTFrameFolder,'/',GTFrameNames{iFrame}];
        iFrameAll = iFrameAll+1;
    end
    fprintf('Found %d frames, video %d/%d\n',iFrameAll-1,iVideo,length(videoNames))
end
fprintf('Found %d frames\n',iFrameAll-1)
end

function generateBatches(frameDirs,saveFolder,nBatches)
global batchSize cropWidth widthNeighbor argumentZoom
% scan for breakpoint
batchesDone = dir(fullfile(saveFolder,'*.mat'));
iBatchStart = 1;
if length(batchesDone) > 0
    iBatchStart = sscanf(batchesDone(end).name,'batch_width%d_size%d_%05d.mat');
    iBatchStart = iBatchStart(end)+2;
end
% start generating
tic;
for iBatch = iBatchStart:nBatches 
    isFrames = randi(length(frameDirs),batchSize,1);
    batchInput = zeros(batchSize,cropWidth,cropWidth,widthNeighbor*3,'uint8');
    batchGT = zeros(batchSize,cropWidth,cropWidth,3,'uint8');
    parfor iPatch = 1:batchSize
        zoom = argumentZoom(unidrnd(4));
        flip = rand>0.5;
        rotate = unidrnd(4);
        
        inputDirs = frameDirs{isFrames(iPatch),1};
        GTDir = frameDirs{isFrames(iPatch),2};
        gtIm = imread(GTDir);
        gtIm = argument(gtIm,zoom,flip,rotate);
        [h,w,c] = size(gtIm);
        patchUncroped = zeros(h,w,c,widthNeighbor);
        patchUncroped(:,:,:,1) = gtIm;
        for iNeighbor = 1:widthNeighbor
            inputIm = imread(inputDirs{iNeighbor});
            patchUncroped(:,:,:,iNeighbor+1) = argument(inputIm,zoom,flip,rotate);
        end
        patchCroped = randomCrop(patchUncroped,cropWidth);
%         montage(patchCroped/255);
%         pause(1)
        batchInput(iPatch,:,:,:) = reshape(patchCroped(:,:,:,2:end),[cropWidth,cropWidth,3*widthNeighbor]);
        batchGT(iPatch,:,:,:) = squeeze(patchCroped(:,:,:,1));
%         imshow(squeeze(batchInput(iPatch,:,:,1:3)));
%         pause(1)
    end
    batchDir = fullfile(saveFolder,sprintf( ...
        'batch_width%d_size%d_%05d.mat',cropWidth,batchSize,iBatch-1));
    disp(['saving to ' batchDir]);
    batchInputTorch = permute(batchInput,[1,4,2,3]);
    batchGTTorch = permute(batchGT,[1,4,2,3]);
    save(batchDir,'batchInputTorch','batchGTTorch','-v6');
    ms = toc*(nBatches-iBatch)/60;
    tic;
    hours = floor(ms/60);
    mins = mod(ms,60);
    fprintf('Generating... %.2f%%, %d hours %.1f minutes left.\n', ...
    iBatch/nBatches*100,hours,mins);
end
end

function imsCrop = randomCrop(ims,cropWidth)
[h,w,~,~] = size(ims);
wRange = w-cropWidth+1;
hRange = h-cropWidth+1;
wStart = randi(wRange);
hStart = randi(hRange);
isHCrop = hStart:hStart+cropWidth-1;
isWCrop = wStart:wStart+cropWidth-1;
imsCrop = ims(isHCrop,isWCrop,:,:);
end

function imArg = argument(im,zoom,flip,rotate)
imArg = imresize(im, zoom);
if flip
    imArg = fliplr(imArg);
end
imArg = rot90(imArg,rotate);
end

File: preprocess/dataArgumentAndCrop.py
import glob
import os
import math
import re
import random
import imageio
import numpy as np
import time
from PIL import Image
import tensorflow as tf
from io import BytesIO

# Parameters
alignments = ['_nowarp','_OF','_homography']
inputDir = '/home/omnisky/Desktop/zc/DeepVideoDeblurring/data'
inputFolderPrefix = 'training_real_all_nostab'
gtDir = '/home/omnisky/Desktop/zc/DeepVideoDeblurring/dataset/quantitative_datasets'

saveDir = '../data'
saveFolderPrefixTrain = 'training_augumented_all_nostab'
saveFolderPrefixValid = 'validating_augumented_all_nostab'

frameExt = '.jpg'
cropWidth = 128
nCrop = 10
nArguments = 2 * 4 * 4
widthNeighbor = 5
batchSize = 64
currSpeed = 0
argumentZoom = [1 / 4, 1 / 3, 1 / 2, 1]

# Scan videos
validset = ['IMG_0030', 'IMG_0049', 'IMG_0021', '720p_240fps_2', 'IMG_0032',
            'IMG_0033', 'IMG_0031', 'IMG_0003', 'IMG_0039', 'IMG_0037']

inputFolders = []
print('Input folders: ')
for alignment in alignments:
    inputFolders.append(os.path.join(inputDir, inputFolderPrefix + alignment))
    print('\t' + inputFolders[-1])

videoNames = []
for inputFolder in inputFolders:
    folders = os.listdir(inputFolder)
    folders.sort()
    # for videoFolder in folders:
    #     print(videoFolder)
    videoNames.append(folders)


# Generate batches
def scanFrames(inputFolder, gtDir, videoNames):
    print('Scanning frames...')
    frameDirs = []
    for iVideo in range(len(videoNames)):
        videoName = videoNames[iVideo]
        inputFrameFolder = os.path.join(inputFolder, videoName)
        inputFrameDirs = glob.glob(os.path.join(inputFrameFolder, 'image_0', '*' + frameExt))
        inputFrameDirs.sort()
        inputFrameNames = [os.path.split(inputFrameDir)[1] for inputFrameDir in inputFrameDirs]
        GTFrameFolder = os.path.join(gtDir, videoName, 'GT')
        GTFrameDirs = glob.glob(os.path.join(GTFrameFolder, '*' + frameExt))
        GTFrameDirs.sort()
        for iFrame in range(len(inputFrameDirs)):
            neighborDirs = []
            for iNeighbor in range(widthNeighbor):
                iFrameNeighbor = iNeighbor - math.floor(widthNeighbor / 2)
                neighborDirs.append(os.path.join(
                    inputFrameFolder,
                    'image_%d' % iFrameNeighbor,
                    inputFrameNames[iFrame]
                ))
            frameDirs.append([neighborDirs, GTFrameDirs[iFrame]])
        print('Found %d frames, video %d/%d' % (len(frameDirs), iVideo + 1, len(videoNames)))
    return frameDirs


def checkDir(dir):
    if not os.path.exists(dir):
        os.makedirs(dir)


def argument(ims, zoom, flip, rotate):
    imsArg = []
    for im in ims:
        s = im.size
        # print(im.size)
        im = im.resize([int(d * zoom) for d in s], Image.BICUBIC)
        # print(im.size)
        if flip:
            im = im.transpose(Image.FLIP_TOP_BOTTOM)
        # print(im.size)
        if rotate == 1:
            im = im.transpose(Image.ROTATE_90)
        if rotate == 2:
            im = im.transpose(Image.ROTATE_180)
        if rotate == 3:
            im = im.transpose(Image.ROTATE_270)
        # print(im.size)
        imsArg.append(im)
    # for im in imsArg:
    #     print(im.size)
    return imsArg


def randomCrop(ims, cropWidth):
    imsCrop = []
    w, h = ims[0].size
    wRange = w - cropWidth
    hRange = h - cropWidth
    wStart = random.randint(0, wRange)
    hStart = random.randint(0, hRange)
    # print(wStart, hStart)
    for im in ims:
        im = im.crop((wStart, hStart, wStart + cropWidth, hStart + cropWidth))
        imsCrop.append(im)
    return imsCrop

from matplotlib import pyplot as plt
def saveBatch(batch, dir):
    print('Saving to ' + dir)
    writer = tf.python_io.TFRecordWriter(dir)
    for patch in batch:
        feature = {}
        for i, im in enumerate(patch):
            if i == 0:
                imName = 'gt'
            else:
                imName = 'input_%d' % (i - 1)
            pngStr = BytesIO()
            im.save(pngStr, format='png')
            pngStr = pngStr.getvalue()
            feature[imName] = tf.train.Feature(bytes_list=tf.train.BytesList(value=[pngStr]))
        example = tf.train.Example(features=tf.train.Features(feature=feature))
        writer.write(example.SerializeToString())
    writer.close()
    pass


def generateBatches(frameDirs, saveFolder, nCrop):
    # batchesDone = glob.glob(os.path.join(saveFolder, '*.tfrecords'))
    # batchesDone = [os.path.split(batcheDone)[1] for batcheDone in batchesDone]
    # batchesDone.sort()
    # if len(batchesDone) > 0:
    #     iBatchStart = int(re.findall('\d{5}', batchesDone[-1])[0])
    # else:
    #     iBatchStart = 0

    # start generating
    batch = []
    iBatch = 0
    nBatches = nCrop * len(frameDirs) * nArguments / batchSize
    tic = time.time()
    for frameDir in frameDirs:
        inputDirs = frameDir[0]
        GTDir = frameDir[1]
        gtIm = Image.open(GTDir)
        inputIms = [Image.open(inputDir) for inputDir in inputDirs]
        for zoom in argumentZoom:
            for flip in range(2):
                for rotate in range(4):
                    gtImArg = argument([gtIm], zoom, flip, rotate)
                    inputImsArg = argument(inputIms, zoom, flip, rotate)
                    for iCrop in range(nCrop):
                        patchCroped = randomCrop(gtImArg + inputImsArg, cropWidth)
                        batch.append(patchCroped)
                        if len(batch) >= batchSize:
                            saveBatch(batch, os.path.join(
                                saveFolder,
                                'batch_width%d_size%d_%05d.tfrecords' % (cropWidth, batchSize, iBatch)
                            ))
                            iBatch += 1
                            batch = []
                            ms = (time.time() - tic) * (nBatches - iBatch) / 60
                            tic = time.time()
                            hours = math.floor(ms / 60)
                            mins = ms % 60
                            print(
                                '%.2f%% %d/%d, %d hours %.1f minutes left.' %
                                ((iBatch) / nBatches * 100, iBatch, nBatches, hours, mins)
                            )

    # for iBatch in range(iBatchStart, int(nBatches)):
    #     batch = []
    #     for iFrame in [random.randint(0, len(frameDirs) - 1) for i in range(batchSize)]:
    #         zoom = random.choice(argumentZoom)
    #         flip = random.randint(0, 1)
    #         rotate = random.randint(0, 3)
    #
    #         inputDirs = frameDirs[iFrame][0]
    #         GTDir = frameDirs[iFrame][1]
    #         gtIm = Image.open(GTDir)
    #         # print(GTDir)
    #         gtIm = argument([gtIm], zoom, flip, rotate)
    #         inputIms = [Image.open(inputDir) for inputDir in inputDirs]
    #         # print(inputDirs)
    #         inputIms = argument(inputIms, zoom, flip, rotate)
    #         patchCroped = randomCrop(gtIm + inputIms, cropWidth)
    #         # print(len(patchCroped))
    #         batch.append(patchCroped)
    #     saveBatch(batch, os.path.join(
    #         saveFolder,
    #         'batch_width%d_size%d_%05d.tfrecords' % (cropWidth, batchSize, iBatch)
    #     ))
    #     ms = (time.time() - tic) * (nBatches - iBatch) / 60
    #     tic = time.time()
    #     hours = math.floor(ms / 60)
    #     mins = ms % 60
    #     print(
    #         '%.2f%% %d/%d, %d hours %.1f minutes left.' %
    #         ((iBatch + 1) / nBatches * 100, iBatch + 1, nBatches, hours, mins)
    #     )


for iAlignment in range(len(alignments)):
    alignment = alignments[iAlignment]

    # scan frames
    trainset = list(set(videoNames[iAlignment]).difference(set(validset)))
    trainset.sort()
    frameDirs = scanFrames(inputFolders[iAlignment], gtDir, trainset)

    # generate batches
    saveAlignFolder = os.path.join(saveDir, saveFolderPrefixTrain + alignment)
    checkDir(saveAlignFolder)
    generateBatches(frameDirs, saveAlignFolder, nCrop)

    # scan validset frames
    validset.sort()
    frameDirs = scanFrames(inputFolders[iAlignment], gtDir, validset)

    # generate validset batches
    saveAlignFolder = os.path.join(saveDir, saveFolderPrefixValid + alignment)
    checkDir(saveAlignFolder)
    generateBatches(frameDirs, saveAlignFolder, nCrop)
